perllib/FixMyStreet/App/Controller/Admin.pm

```diff
@@ -834,80 +834,107 @@ sub categories_for_point : Private {
     if ($template_id eq 'new') {
-        $template = $body->response_templates->new({});
+        $template = $c->stash->{body}->response_templates->new({});
     }
     else {
-        $template = $body->response_templates->find( $template_id )
+        $template = $c->stash->{body}->response_templates->find( $template_id )
             or $c->detach( '/page_error_404_not_found' );
     }
 
+    $c->forward('display_contacts');
+    my @contacts = $template->contacts->all;
+    my @live_contacts = $c->stash->{live_contacts}->all;
+    my %active_contacts = map { $_->id => 1 } @contacts;
+    my @all_contacts = map { {
+        id => $_->id,
+        category => $_->category,
+        active => $active_contacts{$_->id},
+    } } @live_contacts;
+    $c->stash->{contacts} = \@all_contacts;
+
     if ($c->req->method eq 'POST') {
         if ($c->get_param('delete_template') eq _("Delete template")) {
+            $template->contact_response_templates->delete_all;
             $template->delete;
         } else {
             $template->title( $c->get_param('title') );
-            $template->text ( $c->get_param('text') );
+            $template->text( $c->get_param('text') );
+            $template->auto_response( $c->get_param('auto_response') ? 1 : 0 );
             $template->update_or_insert;
+
+            my @live_contact_ids = map { $_->id } @live_contacts;
+            my @new_contact_ids = grep { $c->get_param("contacts[$_]") ? 1 : undef } @live_contact_ids;
+            $template->contact_response_templates->search({
+                contact_id => { '!=' => \@new_contact_ids },
+            })->delete;
+            foreach my $contact_id (@new_contact_ids) {
+                $template->contact_response_templates->find_or_create({
+                    contact_id => $contact_id,
+                });
+            }
         }
 
-        $c->res->redirect( $c->uri_for( 'templates', $body->id ) );
+        $c->res->redirect( $c->uri_for( 'templates', $c->stash->{body}->id ) );
     }
 
     $c->stash->{response_template} = $template;
 
     $c->stash->{template} = 'admin/template_edit.html';
 }
 
+sub load_template_body : Private {
+    my ($self, $c, $body_id) = @_;
 
-sub templates_for_body {
-    my ( $self, $c, $body ) = @_;
+    my $zurich_user = $c->user->from_body && $c->cobrand->moniker eq 'zurich';
+    my $has_permission = $c->user->has_body_permission_to('template_edit') &&
+                         $c->user->from_body->id eq $body_id;
 
-    $c->stash->{body} = $body;
-
-    my @templates = $body->response_templates->search(
-        undef,
-        {
-            order_by => 'title'
-        }
-    );
+    unless ( $c->user->is_superuser || $zurich_user || $has_permission ) {
+        $c->detach( '/page_error_404_not_found' );
+    }
 
-    $c->stash->{response_templates} = \@templates;
+    # Regular users can only view their own body's templates
+    $body_id = $c->user->from_body->id unless $c->user->is_superuser;
```

Comment:
I think it'd be better to redirect to the right page if they've come to a page for a different body (I think that's what this line would mean?).